apiservice/apiservice.go

```diff
@@ -0,0 +1,275 @@
+)
+
+func NewFlowCachedAccessAPI(accessNodeAddressAndPort flow.IdentityList, timeout time.Duration) (*FlowCachedAccessAPI, error) {
+	ret := &FlowCachedAccessAPI{}
+	ret.timeout = timeout
+	ret.upstream = make([]access.AccessAPIClient, accessNodeAddressAndPort.Count())
+	ret.connections = make([]*grpc.ClientConn, accessNodeAddressAndPort.Count())
+	for i, identity := range accessNodeAddressAndPort {
+		// Store the client setup parameters such as address, public, key and timeout, so that
+		// we can refresh the API on connection loss
+		ret.ids[i] = identity
+
+		// We fail on any single error on startup, so that
+		// we identify bootstrapping errors early
+		err := ret.updateClient(i)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	ret.roundRobin = 0
+	return ret, nil
+}
+
+type FlowCachedAccessAPI struct {
+	access.AccessAPIServer
+	lock        sync.Mutex
+	roundRobin  int
+	ids         flow.IdentityList
+	upstream    []access.AccessAPIClient
+	connections []*grpc.ClientConn
+	timeout     time.Duration
+}
+
+func (h *FlowCachedAccessAPI) SetLocalAPI(local access.AccessAPIServer) {
+	h.AccessAPIServer = local
+}
+
+func (h *FlowCachedAccessAPI) updateClient(i int) error {
+	timeout := h.timeout
+
+	if h.connections[i] == nil || h.connections[i].GetState() != connectivity.Ready {
+		identity := h.ids[i]
+		if identity.NetworkPubKey == nil {
+			clientRPCConnection, err := grpc.Dial(
+				identity.Address,
+				grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(grpcutils.DefaultMaxMsgSize)),
+				grpc.WithInsecure(), //nolint:staticcheck
+				backend.WithClientUnaryInterceptor(timeout))
+			if err != nil {
+				return err
+			}
+
+			h.connections[i] = clientRPCConnection
+			h.upstream[i] = access.NewAccessAPIClient(clientRPCConnection)
+		} else {
+			tlsConfig, err := grpcutils.DefaultClientTLSConfig(identity.NetworkPubKey)
+			if err != nil {
+				return fmt.Errorf("failed to get default TLS client config using public flow networking key %s %w", identity.NetworkPubKey.String(), err)
+			}
+
+			clientRPCConnection, err := grpc.Dial(
+				identity.Address,
+				grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(grpcutils.DefaultMaxMsgSize)),
+				grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig)),
+				backend.WithClientUnaryInterceptor(timeout))
+			if err != nil {
+				return err
+			}
+
+			h.connections[i] = clientRPCConnection
+			h.upstream[i] = access.NewAccessAPIClient(clientRPCConnection)
+		}
+	}
+
+	return nil
+}
+
+func (h *FlowCachedAccessAPI) client() (access.AccessAPIClient, error) {
+	if h.upstream == nil || len(h.upstream) == 0 {
+		return nil, status.Errorf(codes.Unimplemented, "method not implemented")
+	}
+
+	h.lock.Lock()
+	defer h.lock.Unlock()
+
+	var err error
+	for i := 0; i < 3; i++ {
+		h.roundRobin++
+		h.roundRobin = h.roundRobin % len(h.upstream)
+		err = h.updateClient(h.roundRobin)
+		if err == nil {
+			return h.upstream[h.roundRobin], nil
+		}
+	}
+
+	return nil, err
```

Comment:
should the error be converted to include a gRPC status code?

Reply:
Addressed: https://github.com/GetElastech/flow-go/commit/53c8a6b4338a3084b1f7cdbcf127a074b6ee3607